File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-24.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
    mergeKLists(lists) { 
        if (lists.length === 0) return null;
        return this.rec(lists, 0, lists.length - 1)
    } 


    rec(lists, l, r) { 
        if (l === r) {
            return lists[l]
        }

        let mid = Math.floor((r + l) / 2) 
        let left = this.rec(lists, l, mid) 
        let right = this.rec(lists, mid + 1, r) 
        return this.merge(left, right)
    }


    merge(l1, l2) {
        let dummy = new ListNode()  
        let node = dummy

        while(l1 && l2) {
            if (l1.val > l2.val) {
                node.next = l2 
                l2 = l2.next
            } else {
                node.next = l1 
                l1 = l1.next
            } 

            node = node.next
        } 

        node.next = l1 ? l1 : l2 

        return dummy.next
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        dummy = ListNode(0, head) 

        fast = dummy 
        slow = dummy 

        while fast and fast.next: 
            fast = fast.next.next 
            slow = slow.next 

        h2 = slow.next 
        slow.next = None 

        prev = None 
        while h2: 
            temp = h2.next 
            h2.next = prev 
            prev = h2 
            h2 = temp 

        h1 = head 
        h2 = prev 

        while h1 and h2: 
            t1 = h1.next 
            t2 = h2.next 
            h1.next = h2 
            h2.next = t1 
            h1 = t1 
            h2 = t2 